httpbp/errors.go

```diff
@@ -0,0 +1,237 @@
+
+	// Optional map of invalid fields to error messages, for use with errors where
+	// the client sent invalid input (400).
+	//
+	// This allows servers to return validation information for multiple
+	// fields at once. Ex:
+	// {
+	//   "foo.name": "This field is required.",
+	//   "bar.id": "This field is required.",
+	// }
+	Details map[string]string `json:"details,omitempty"`
+
+	code int
+	raw  string
+}
+
+// SetRawResponse is used to set the respose to return when using a Raw content
+// writer.
+//
+// This is ignored when using any other content writer.
+func (r *ErrorResponse) SetRawResponse(raw string) {
+	r.raw = raw
+}
+
+// String fufills the fmt.Stringer interface by returning b.RawResponse if it is
+// set, otherwise returns b.Reason.
+//
+// This will be used if you are using returning this as a Raw error.
+func (r ErrorResponse) String() string {
+	if r.raw != "" {
+		return r.raw
+	}
+	return r.Reason
+}
+
+// TemplateName fufills the httpbp.HTMLBody interface by returning the name of
+// the template to use to render the error page when using an HTML response.
+//
+// The format of the template name is "httpbp/errors/{response_code}".
+func (r ErrorResponse) TemplateName() string {
+	return fmt.Sprintf("httpbp/errors/%d", r.code)
+}
+
+// NewErrorResponse returns a new ErrorResponse with the given inputs.
+//
+// This can be used to create custom NewErrorResponses, but it is encouraged that
+// developers use the standard ones provided by httpbp rather than creating them
+// directly.
+func NewErrorResponse(code int, reason string, explanation string) *ErrorResponse {
+	return &ErrorResponse{
+		code:        code,
+		Reason:      reason,
+		Explanation: explanation,
+		Details:     make(map[string]string),
+	}
+}
+
+var (
+	_ HTMLBody     = ErrorResponse{}
+	_ HTMLBody     = (*ErrorResponse)(nil)
+	_ fmt.Stringer = ErrorResponse{}
+	_ fmt.Stringer = (*ErrorResponse)(nil)
+)
+
+// JSONError returns the given error as an HTTPError that will write JSON.
+func JSONError(resp *ErrorResponse, cause error) HTTPError {
+	body := ErrorResponseJSONWrapper{Success: false, Error: resp}
+	return newHTTPError(resp.code, body, cause, JSONContentWriter())
+}
+
+// HTMLError returns the given error as an HTTPError that will write HTML.
+func HTMLError(resp *ErrorResponse, cause error, t *template.Template) HTTPError {
+	return newHTTPError(resp.code, resp, cause, HTMLContentWriter(t))
+}
+
+// RawError returns the given error as an HTTPError that will write a raw
+// response of the given content-type.
+func RawError(resp *ErrorResponse, cause error, contentType string) HTTPError {
+	return newHTTPError(resp.code, resp, cause, RawContentWriter(contentType))
+}
+
+// RetryableError can be used to wrap an ErrorResponse and communicate to the
+// caller that the request can be retried after the given duration.
+func RetryableError(resp ErrorResponse, w http.ResponseWriter, retryAfter time.Duration) ErrorResponse {
```

Comment:
🔕 Although semantically this API is related to an error, but technically it didn't touch the `resp` passed in and always return it as-is. This means this can be written completely as a helper around `http.ResponseWriter`, without involving `ErrorResponse`.

Reply:
Ya, maybe this is _too_ cute?  My thinking was it makes for nice looking code to be able to do something like:

```go
return httpbp.JSONError(
    httpbp.RetryableError(httpbp.ServiceUnavailable(), w, time.Hour), errors.New("Downtime"),
)
```

Reply:
hmm that's a reasonable use case, maybe just add to comment to explain it? also probably extract the version don't require `resp` as well :)

Reply:
_technically_ you can still do:

```
httpbp.WriteRetryableError(e, time.Hour)
return httpbp.JSONError(httpbp.ServiceUnavailable(), errors.New("downtime"))
```

but I can see that language sugar still be useful :)

Reply:
I'll leave it like this for now and we can mull over it while I grind out tests/errors for each response code 😅

Reply:
I think I found a nicer way to do it actually:

```go
// Retryable communicates to the  caller that the request may be retried after
// the given duration.
func (r *ErrorResponse) Retryable(w http.ResponseWriter, retryAfter time.Duration) *ErrorResponse {
	after := strconv.FormatInt(int64(retryAfter/time.Second), 10)
	w.Header().Set(RetryAfterHeader, after)
	return r
}
```

So it's now:

```go
return httpbp.JSONError(
    httpbp.ServiceUnavailable().Retryable(w, time.Hour), errors.New("Downtime"),
)
```

Reply:
Nice. The new approach makes a lot of sense :+1: